Source/Core/VideoCommon/OpcodeDecoding.cpp

```diff
@@ -81,202 +41,229 @@ void Init()
 
-        // 4 GPU ticks per vertex, 3 CPU ticks per GPU tick
-        total_cycles += num_vertices * 4 * 3 + 6;
-      }
-      else
-      {
-        if (!s_is_fifo_error_seen)
-          CommandProcessor::HandleUnknownOpcode(cmd_byte, opcode_start, is_preprocess);
-        ERROR_LOG_FMT(VIDEO, "FIFO: Unknown Opcode({:#04x} @ {}, preprocessing = {})", cmd_byte,
-                      fmt::ptr(opcode_start), is_preprocess ? "yes" : "no");
-        s_is_fifo_error_seen = true;
-        total_cycles += 1;
+          // un-swap
+          g_stats.SwapDL();
+        }
       }
-      break;
+
+      m_in_display_list = false;
+    }
+  }
+  DOLPHIN_FORCE_INLINE void OnNop(u32 count) override
+  {
+    m_cycles += 6 * count;  // Hm, this means that we scan over nop streams pretty slowly...
+  }
+  DOLPHIN_FORCE_INLINE void OnUnknown(u8 opcode, const u8* data) override
+  {
+    if (static_cast<Opcode>(opcode) == Opcode::GX_UNKNOWN_RESET)
+    {
+      // Datel software uses this command
+      m_cycles += 6;
+      DEBUG_LOG_FMT(VIDEO, "GX Reset?");
+    }
+    else if (static_cast<Opcode>(opcode) == Opcode::GX_CMD_UNKNOWN_METRICS)
+    {
+      // zelda 4 swords calls it and checks the metrics registers after that
+      m_cycles += 6;
+      DEBUG_LOG_FMT(VIDEO, "GX 0x44");
+    }
+    else if (static_cast<Opcode>(opcode) == Opcode::GX_CMD_INVL_VC)
+    {
+      // Invalidate Vertex Cache
+      m_cycles += 6;
+      DEBUG_LOG_FMT(VIDEO, "Invalidate (vertex cache?)");
     }
+    else
+    {
+      if (!s_is_fifo_error_seen)
+        CommandProcessor::HandleUnknownOpcode(opcode, data, is_preprocess);
+      ERROR_LOG_FMT(VIDEO, "FIFO: Unknown Opcode({:#04x} @ {}, preprocessing = {})", opcode,
+                    fmt::ptr(data), is_preprocess ? "yes" : "no");
+      s_is_fifo_error_seen = true;
+      m_cycles += 1;
+    }
+  }
 
-    // Display lists get added directly into the FIFO stream
+  DOLPHIN_FORCE_INLINE void OnCommand(const u8* data, u32 size) override
+  {
+    ASSERT(size >= 1);
     if constexpr (!is_preprocess)
     {
-      if (g_record_fifo_data && cmd_byte != GX_CMD_CALL_DL)
+      // Display lists get added directly into the FIFO stream since this same callback is used to
+      // process them.
+      if (g_record_fifo_data && static_cast<Opcode>(data[0]) != Opcode::GX_CMD_CALL_DL)
       {
-        const u8* const opcode_end = src.GetPointer();
-        FifoRecorder::GetInstance().WriteGPCommand(opcode_start, u32(opcode_end - opcode_start));
+        FifoRecorder::GetInstance().WriteGPCommand(data, size);
       }
     }
   }
+
+  DOLPHIN_FORCE_INLINE CPState& GetCPState() override
+  {
+    if constexpr (is_preprocess)
+      return g_preprocess_cp_state;
+    else
+      return g_main_cp_state;
+  }
+
+  u32 m_cycles;
+  bool m_in_display_list;
+};
+
+// This code is very hot (as everything in the FIFO passes through it and it needs to work
+// realtime). Thus, some hackery is used above to inline a lot of code into it.
```

Comment:
```suggestion
// This code is very hot; as everything in the FIFO passes through it and it needs to work
// in realtime. Thus, some hackery is used above to inline a lot of code into it.
```
Shouldn't this comment be next to the relevant code _above_ anyway?

Reply:
I've removed the comment.  It made more sense before when there were inline versions of `Run` and `RunCommand` right above in `OpcodeDecoding.cpp`, but I've since moved them to the header so that _all_ uses of them can be inlined.  The current, more relevant comment is [here](https://github.com/Pokechu22/dolphin/blob/8e468820a5ef6fa5d013872e030bb9cca989862f/Source/Core/VideoCommon/OpcodeDecoding.h#L66-L69) (though perhaps I should add a bit more info about inlining there).